transactions/update-contract/2024/apr-24-upgrade-random-history/RandomBeaconHistory.cdc

```diff
@@ -0,0 +1,344 @@
+/// the safe Random Beacon.
+///
+/// Read the full FLIP here: https://github.com/onflow/flips/pull/123
+///
+access(all) contract RandomBeaconHistory {
+
+    /// The height at which the first source of randomness was recorded
+    access(contract) var lowestHeight: UInt64?
+    /// Sequence of random sources recorded by the Heartbeat, stored as an array over a mapping to reduce storage
+    access(contract) let randomSourceHistory: [[UInt8]]
+
+    /// The path of the Heartbeat resource in the deployment account
+    access(all) let HeartbeatStoragePath: StoragePath
+
+    // Event emitted when missing SoRs from past heartbeats are detected and will be backfilled:
+    //  - `blockHeight` is the height where the gap is detected
+    //  - `gapStartHeight` is the height of the first missing entry detected
+    access(all) event RandomHistoryMissing(blockHeight: UInt64, gapStartHeight: UInt64)
+
+    // Event emitted when missing SoRs are backfilled on the current heartbeat:
+    //  - `blockHeight` is the height where the backfill happened, it also defines the SoR used to backfill
+    //  - `gapStartHeight` is the height of the first backfilled entry
+    //  - `count` is the number of backfilled entries
+    // Note that in very rare cases, the backfilled gap may not be contiguous. This event does not
+    // fully define the backfilled entries in this case.
+    access(all) event RandomHistoryBackfilled(blockHeight: UInt64, gapStartHeight: UInt64, count: UInt64)
+
+    /* --- Hearbeat --- */
+    //
+    /// The Heartbeat resource containing each block's source of randomness in sequence
+    ///
+    access(all) resource Heartbeat {
+
+        /// Callable by owner of the Heartbeat resource, Flow Service Account, records the provided random source
+        ///
+        /// @param randomSourceHistory The random source to record
+        ///
+        /// The Flow protocol makes sure to call this function once per block as a system call. The transaction
+        /// comes at the end of each block so that the current block's entry becomes available only in the child
+        /// block.
+        ///
+        access(all) fun heartbeat(randomSourceHistory: [UInt8]) {
+
+            assert(
+                // random source must be at least 128 bits
+                randomSourceHistory.length >= 128 / 8,
+                message: "Random source must be at least 128 bits"
+            )
+
+            let currentBlockHeight = getCurrentBlock().height
+            // init lowestBlockHeight if it is not set yet
+            if RandomBeaconHistory.lowestHeight == nil {
+                RandomBeaconHistory.lowestHeight = currentBlockHeight
+            }
+            var maybeBackfiller = RandomBeaconHistory.borrowBackfiller()
+            // Create & save Backfiller if it is not saved yet
+            if maybeBackfiller == nil {
+                RandomBeaconHistory.account.storage.save(<-create Backfiller(), to: /storage/randomBeaconHistoryBackfiller)
```

Comment:
What network is this going to? I'm realizing this is Cadence 1.0 syntax and won't work on Testnet or Mainnet

Reply:
oops, I copied it from the wrong branch. I'll update it with the correct one

Reply:
thanks for catching that!

Reply:
updated